### DISCUSSION TITLE
string type converted automatically to array.

### DISCUSSION THREAD

**Discussion by fjanisze** [nlohmann/json#2614] [Q&A] [answered]
Hello,

I've some strange behaviors while using json.hpp in my C++ application:

Given a map of json items:

```cpp
std::map< std::string, nlohmann::json>;
```

Which is filled by such a lambda:

```cpp
 auto add_option = [this](
            const std::string& name,
            nlohmann::json value
    ) {
        dbg("Adding option %s:%s (%s) to %s",
            name.c_str(),
            value.dump().c_str(),
            value.type_name(),
            id_.c_str());
        opts_[ name ] = value; //HERE IS FILLED
    };
```

That for example for '_type' generate the output "Adding option _type:"ux-socket" (string) to socket-input", while attempting to access the data with a function like this:

```cpp
nlohmann::json
get_option(
    const std::string& name
)
{
    auto it = opts_.find( name );
    if( it != opts_.end() ) {
        logp("RETURNING %s",
             it->second.type_name()); //THIS IS JUST FOR DEBUG
        return it->second;
    }
    return nlohmann::json::value_t::null;
}
```

For name == _type I can see the RETURNING print which confirm the type if 'string', but then if I attempt to use the function:

```cpp
 auto value{ get_option( name ) }; //SEE HERE
    if( !value.is_string() ) {  //value IS NOT A STRING ANYMORE HERE..
        dbg("The option %s (%s : %s) is not a string!",
            name.c_str(),
            value.type_name(),
            value.dump().c_str());
    }
```

_type is not a string anymore, and now is an array, I get the following output: "The option _type (array : ["ux-socket"]) is not a string!" how did that happen that my string became an array? in get_option it's a string, when i call get_option it became an array.. is that ok?

Does my explanation make sense? Further details needed perhaps?

Thanks

**Comment by nlohmann** [comment, marked as answer]
The problem are the braces here:

```cpp
auto value{ get_option( name ) };
```

These braces are interpreted as an array (see #2583 for the latest discussion on why this was a rather bug-prone design decision). So whatever value is retuned, it will be put into an array.

You can fix this by changing this to

```cpp
auto value = get_option( name );
```

and everything should be fine.
